File: uva online judge/10611 - Playboy Chimp.py
import math

def shortest(heights, i, j, h):
    k = (i + j) // 2
    if i == j:
        return i
    if heights[k] > h:
        return shortest(heights, i, k, h)
    return shortest(heights, k + 1, j, h)

def tallest(heights, i, j, h):
    k = math.ceil((i + j) / 2)
    if i == j:
        return i
    if heights[k] >= h:
        return tallest(heights, i, k - 1, h)
    return tallest(heights, k, j, h)

if __name__ == '__main__':
    n = int(input())
    heights = [int(i) for i in input().split()]
    q = int(input())
    for h in (int(i) for i in input().split()):
        tallest_i = tallest(heights, 0, n - 1, h)
        shortest_i = shortest(heights, 0, n - 1, h)
        print(heights[tallest_i] if heights[tallest_i] < h  else 'X', heights[shortest_i] if heights[shortest_i] > h else 'X')

File: uva online judge/10611 - Playboy Chimp.c
#include <stdio.h>
#include <math.h>

int heights[50000], queries[25000];

int smallest(int heights[], int i, int j, int h){
    int k;
    if(i == j)
        return i;
    k = (i + j) / 2;
    if(heights[k] > h)
        return smallest(heights, i, k, h);
    return smallest(heights, k + 1, j, h);
}

int tallest(int heights[], int i, int j, int h){
    int k;
    if(i == j)
        return i;
    k = ceil((i + j) / 2.0);
    if(heights[k] >= h)
        return tallest(heights, i, k - 1, h);
    return tallest(heights, k, j, h);
}

void read_test_case(int *n, int *q, int heights[], int queries[]){
    int i;
    scanf("%d", n);
    for(i = 0; i < *n; i++)
        scanf("%d", &heights[i]);
    scanf("%d", q);
    for(i = 0; i < *q; i++)
        scanf("%d", &queries[i]);
}

int main(){
    int n, q, i, smallest_i, tallest_i;
    read_test_case(&n, &q, heights, queries);
    for(i = 0; i < q; i++){
        smallest_i = smallest(heights, 0, n - 1, queries[i]);
        tallest_i = tallest(heights, 0, n - 1, queries[i]);
        if(heights[tallest_i] < queries[i])
            printf("%d", heights[tallest_i]);
        else
            printf("X");
        if(heights[smallest_i] > queries[i])
            printf(" %d\n", heights[smallest_i]);
        else
            printf(" X\n");
    }
    return 0;
}
